In [201]:
%reset

Once deleted, variables cannot be recovered. Proceed (y/[n])?  y


In [202]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns

In [203]:
X1=pd.Series([1,3,4,7,9,9],name="X1")
X2=pd.Series([9,9,6,3,1,2],name="X2")
Y=pd.Series([3,5,6,8,7,10],name="Y")

In [204]:
from statsmodels.formula.api import ols

DF=pd.concat([X1,X2,Y],axis='columns')
print(DF)

   X1  X2   Y
0   1   9   3
1   3   9   5
2   4   6   6
3   7   3   8
4   9   1   7
5   9   2  10


In [205]:
model1=ols('Y~X1+X2',data=DF)
results1=model1.fit()
print(results1.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.835
Model:                            OLS   Adj. R-squared:                  0.726
Method:                 Least Squares   F-statistic:                     7.611
Date:                Wed, 20 Jul 2022   Prob (F-statistic):             0.0668
Time:                        15:12:24   Log-Likelihood:                -7.8795
No. Observations:                   6   AIC:                             21.76
Df Residuals:                       3   BIC:                             21.13
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.6927      7.363     -0.230      0.8

C:\Users\carol\anaconda3\lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 6 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


In [206]:
np.mean(DF["X1"])

5.5

In [207]:
np.mean(DF["X2"])

5.0

In [208]:
# Matrix mean centered X values

XMatrix=np.array([[-4.5,4],[-2.5,4],[-1.5,1],[1.5,-2],[3.5,-4],[3.5,-3]])
XMatrix

array([[-4.5,  4. ],
       [-2.5,  4. ],
       [-1.5,  1. ],
       [ 1.5, -2. ],
       [ 3.5, -4. ],
       [ 3.5, -3. ]])

In [209]:
Y1=np.array([[-3.5],[-1.5],[-0.5],[1.5],[0.5],[3.5]])
Y1

array([[-3.5],
       [-1.5],
       [-0.5],
       [ 1.5],
       [ 0.5],
       [ 3.5]])

we will use matrix form for working with linear algebra equations

b= ((X transpose X) inverse) * (X transpose y)

numpy linear algebra https://numpy.org/doc/stable/reference/routines.linalg.html 

First find ((X transpose X) inverse)

In [210]:
(XMatrix.T)

# 2 by 6 matrix, 2 rows and 6 coulms 

array([[-4.5, -2.5, -1.5,  1.5,  3.5,  3.5],
       [ 4. ,  4. ,  1. , -2. , -4. , -3. ]])

In [211]:
XMatrix

# 6 by 2 matrix
# for matrix multiplication to be defined, 
# the number of columns in the first matrix must be equal to the number of rows in the second matrix.

array([[-4.5,  4. ],
       [-2.5,  4. ],
       [-1.5,  1. ],
       [ 1.5, -2. ],
       [ 3.5, -4. ],
       [ 3.5, -3. ]])

In [212]:
(XMatrix.T.dot(XMatrix))

array([[ 55.5, -57. ],
       [-57. ,  62. ]])

In [213]:
# This is the scaled version of the covariance of the X matrix
# SS of X1 and X2 on main diagonal and sum of cross products on the off diagonal: 
# strength of the relationship between X variables

# In python the covariance matrix can be found using np.cov, 
# where the variances are the main diagonal and the covariance is the other diagonal. 
# The covariance matrix provides the covariance between the pairs of elements in the vectors of data.  

np.cov(X1,X2)

array([[ 11.1, -11.4],
       [-11.4,  12.4]])

In [214]:
XtXinv=np.linalg.inv(XMatrix.T.dot(XMatrix))

Now find (X transpose y)

This is the covariance of the X variables with y

In [215]:
(XMatrix.T)

# 2 by 6 matrix

array([[-4.5, -2.5, -1.5,  1.5,  3.5,  3.5],
       [ 4. ,  4. ,  1. , -2. , -4. , -3. ]])

In [216]:
Y1

# 6 by 1 matrix

array([[-3.5],
       [-1.5],
       [-0.5],
       [ 1.5],
       [ 0.5],
       [ 3.5]])

In [217]:
XtY=(XMatrix.T).dot(Y1)

Finally, multiply the reuslts

b= ((X transpose X) inverse) * (X transpose y)

In [218]:
XtXinv.dot(XtY)

array([[1.09895833],
       [0.4296875 ]])

In [219]:
np.linalg.inv(XMatrix.T.dot(XMatrix)).dot(XMatrix.T).dot(Y1)

array([[1.09895833],
       [0.4296875 ]])

this is the coefficient matrix with b1 = 1.099 and b2 = 0.430

to get back the y-intercept b0, you can use the equation for the regression equation

y=b0 +b1X1 +b2X2

b0 = y(mean) - X1(mean)*b1 - X2(mean)*b2

In [220]:
np.mean(Y)-np.mean(X1)*1.09895833-np.mean(X2)*0.4296833

-1.6926873150000006

You can check with OLS

In [221]:
import statsmodels.api as sm

model2=sm.OLS(Y1,XMatrix)
results2=model1.fit()
print(results2.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.835
Model:                            OLS   Adj. R-squared:                  0.726
Method:                 Least Squares   F-statistic:                     7.611
Date:                Wed, 20 Jul 2022   Prob (F-statistic):             0.0668
Time:                        15:12:27   Log-Likelihood:                -7.8795
No. Observations:                   6   AIC:                             21.76
Df Residuals:                       3   BIC:                             21.13
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.6927      7.363     -0.230      0.8

C:\Users\carol\anaconda3\lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 6 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


You can use the matrices with the column of 1's directly using the OLS call.

In [222]:
XMatrix1=np.array([[1,1,9],[1,3,9],[1,4,6],[1,7,3],[1,9,1],[1,9,2]])
XMatrix1

array([[1, 1, 9],
       [1, 3, 9],
       [1, 4, 6],
       [1, 7, 3],
       [1, 9, 1],
       [1, 9, 2]])

In [223]:
Y2=([[3],[5],[6],[8],[7],[10]])

In [224]:
import statsmodels.api as sm

model3=sm.OLS(Y2,XMatrix1)
results3=model3.fit()
print(results3.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.835
Model:                            OLS   Adj. R-squared:                  0.726
Method:                 Least Squares   F-statistic:                     7.611
Date:                Wed, 20 Jul 2022   Prob (F-statistic):             0.0668
Time:                        15:12:29   Log-Likelihood:                -7.8795
No. Observations:                   6   AIC:                             21.76
Df Residuals:                       3   BIC:                             21.13
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.6927      7.363     -0.230      0.8

C:\Users\carol\anaconda3\lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 6 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


You can use a DF as usual directly in ols

In [225]:
XMatrix5=pd.DataFrame([[1,9,3],[3,9,5],[4,6,6],[7,3,8],[9,1,7],[9,2,10]],columns=["X1","X2","y"])
XMatrix5

X1  X2   y
0   1   9   3
1   3   9   5
2   4   6   6
3   7   3   8
4   9   1   7
5   9   2  10

In [227]:
from statsmodels.formula.api import ols
import statsmodels.formula.api as sm

model4=sm.ols('y~X1+X2',data=XMatrix5)
results4=model4.fit()
print(results4.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.835
Model:                            OLS   Adj. R-squared:                  0.726
Method:                 Least Squares   F-statistic:                     7.611
Date:                Wed, 20 Jul 2022   Prob (F-statistic):             0.0668
Time:                        15:13:55   Log-Likelihood:                -7.8795
No. Observations:                   6   AIC:                             21.76
Df Residuals:                       3   BIC:                             21.13
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.6927      7.363     -0.230      0.8

C:\Users\carol\anaconda3\lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 6 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "
